include/cuda/memory_resource

```diff
@@ -0,0 +1,309 @@
+                                 _PropertyFns... __property_fn_) noexcept
+      : _Property_vtable<_Properties>(__property_fn_)...,
+        __alloc_fn(__alloc_fn_), __dealloc_fn(__dealloc_fn_),
+        __equal_fn(__equal_fn_) {}
+
+  template <class _Resource>
+  inline __host__ __device__ static constexpr _Resource_vtable
+  _Create() noexcept {
+    return {
+        &_Resource_vtable_builder::_Alloc<_Resource>,
+        &_Resource_vtable_builder::_Dealloc<_Resource>,
+        &_Resource_vtable_builder::_Equal<_Resource>,
+        &_Resource_vtable_builder::_Get_property<_Resource, _Properties>...};
+  }
+
+  template <class _Property>
+  inline __host__ __device__ constexpr typename _Property::value_type
+  _Get_property(void *__object, _Property) const noexcept {
+    return _Property_vtable<_Property>::__property_fn(__object);
+  }
+};
+
+template <class... _Properties> struct _Filtered;
+
+template <bool _IsUniqueProperty> struct _Property_filter {
+  template <class _Property, class... _Properties>
+  using _Filtered_properties = typename _Filtered<
+      _Properties...>::_Filtered_vtable::template _Append_property<_Property>;
+};
+template <> struct _Property_filter<false> {
+  template <class _Property, class... _Properties>
+  using _Filtered_properties =
+      typename _Filtered<_Properties...>::_Filtered_vtable;
+};
+
+template <class _Property, class... _Properties>
+struct _Filtered<_Property, _Properties...> {
+  // clang-format off
+    using _Filtered_vtable = typename _Property_filter<property_with_value_v<_Property> //
+        && !_CUDA_VSTD::_One_of<_Property, _Properties...>>::template _Filtered_properties<_Property, _Properties...>;
+  // clang-format on
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery, _Property, _Properties...>;
+
+  using _VTable = _Resource_vtable<_Property, _Properties...>;
+};
+
+template <> struct _Filtered<> {
+  using _Filtered_vtable = _Filtered<>;
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery>;
+
+  using _VTable = _Resource_vtable<>;
+};
+
+template <class... _Properties>
+using _Filtered_vtable =
+    typename _Filtered<_Properties...>::_Filtered_vtable::_VTable;
+
+// clang-format off
+template <class _Resource, class... _Properties>
+__device__ _LIBCUDACXX_INLINE_VAR constexpr _Filtered_vtable<_Properties...> __resource_vtable =
+    _Filtered_vtable<_Properties...>::template _Create<_Resource>();
+// clang-format on
+
+template <class... _Properties> class any_resource_ref {
+private:
+  template <class... _OtherProperties> friend class any_resource_ref;
+
+  void *__object = nullptr;
+  const _Filtered_vtable<_Properties...> *__vtable = nullptr;
+
+public:
+  // clang-format off
+    _LIBCUDACXX_TEMPLATE(class _Resource)
+        (requires resource_with<_Resource, _Properties...>)
+    inline __host__ __device__
+        any_resource_ref(_Resource& __res) noexcept
```

Comment:
Note to self: Whitespace issues. 

That said, the whole formatting is a bloody mess due to the concept emulation